compiler/src/dotty/tools/backend/jvm/GenBCode.scala

```diff
@@ -46,14 +47,37 @@ class GenBCode extends Phase {
     superCallsMap.put(sym, old + calls)
   }
 
-  def outputDir(implicit ctx: Context): AbstractFile =
-    new PlainDirectory(ctx.settings.outputDir.value)
+  def outputDir(implicit ctx: Context): AbstractFile = {
+    val path = ctx.settings.outputDir.value
+    if (path.isDirectory) new PlainDirectory(Directory(path))
+    else new PlainFile(path)
+  }
+
+  private[this] var classOutput: AbstractFile = _
 
   def run(implicit ctx: Context): Unit = {
     new GenBCodePipeline(entryPoints.toList,
-        new DottyBackendInterface(outputDir, superCallsMap.toMap)(ctx))(ctx).run(ctx.compilationUnit.tpdTree)
+        new DottyBackendInterface(classOutput, superCallsMap.toMap)(ctx))(ctx).run(ctx.compilationUnit.tpdTree)
     entryPoints.clear()
   }
+
+  override def runOn(units: List[CompilationUnit])(implicit ctx: Context) = {
+    val output = outputDir
+    if (output.isDirectory) {
+      classOutput = output
+      val res = super.runOn(units)
+      classOutput = null
+      res
+    } else {
+      assert(output.hasExtension("jar"))
+      classOutput = new PlainDirectory(Path(Path(output.file).parent + "/tmp-jar-" + System.currentTimeMillis().toHexString).createDirectory())
```

Comment:
scalac has code to write directly to a JAR instead of using a temporary directory which should be faster, maybe we can reuse that: https://github.com/scala/scala/blob/d41c97f13ea2348ec6f59c24ae098d83a96f30dc/src/compiler/scala/tools/nsc/backend/jvm/ClassfileWriter.scala#L23 Although apparently in Java >= 7 there's already an API dedicated to this: https://github.com/sbt/zinc/issues/305#issuecomment-308093743

Reply:
The `ClassfileWriter` came in the large refactoring made in scala/scala#6012. You said previously in #3113 that you would rather not port this one.

Reply:
The Java 7 API does not allow direct access to the files in the jar.

`Once you have an instance of a zip file system, you can invoke the methods of the java.nio.file.FileSystem and java.nio.file.Path classes to perform operations such as copying, moving, and renaming files, as well as modifying file attributes.` [quote link](https://docs.oracle.com/javase/7/docs/technotes/guides/io/fsp/zipfilesystemprovider.html)

Reply:
Can't you do something like:
```scala
def outputDir(implicit ctx: Context): AbstractFile = {
  if (isJar) {
    val jar = Paths.get(pathTojar)
    val fs = FileSystems.newFileSystem(jar)
    new PlainNioFile(fs.getPath("/"))
  }
  ...
}
```

Reply:
I will try

Reply:
It worked with the `java.nio.file. FileSystem` jar.